velox/functions/prestosql/Sequence.cpp

```diff
@@ -0,0 +1,152 @@
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+class SequenceFunction : public exec::VectorFunction {
+ public:
+  static constexpr int32_t kMaxResultEntries = 10'000;
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VectorPtr localResult = applyNumber(rows, args, outputType, context);
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+ private:
+  static VectorPtr applyNumber(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) {
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto startVector = decodedArgs.at(0);
+    auto stopVector = decodedArgs.at(1);
+    DecodedVector* stepVector = nullptr;
+    if (args.size() == 3) {
+      stepVector = decodedArgs.at(2);
+    }
+
+    const auto numRows = rows.end();
+    auto pool = context.pool();
+    vector_size_t numElements = 0;
+
+    BufferPtr sizes = allocateSizes(numRows, pool);
+    BufferPtr offsets = allocateOffsets(numRows, pool);
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      auto start = startVector->valueAt<int64_t>(row);
+      auto stop = stopVector->valueAt<int64_t>(row);
+      const int64_t step = (stepVector == nullptr)
+          ? (stop >= start ? 1 : -1)
+          : stepVector->valueAt<int64_t>(row);
+      checkArguments(start, stop, step);
+      rawSizes[row] = (stop - start) / step + 1;
+      numElements += rawSizes[row];
+    });
+
+    VectorPtr elements = BaseVector::create(BIGINT(), numElements, pool);
+    auto rawElements = elements->asFlatVector<int64_t>()->mutableRawValues();
+
+    vector_size_t elementsOffset = 0;
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      const auto numElements = rawSizes[row];
+      if (numElements) {
+        rawOffsets[row] = elementsOffset;
+        writeToElements(
+            rawElements,
+            elementsOffset,
+            startVector,
+            stopVector,
+            stepVector,
+            row);
+        elementsOffset += rawSizes[row];
+      }
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool, outputType, nullptr, numRows, offsets, sizes, elements);
+  }
+
+  static void checkArguments(int64_t start, int64_t stop, int64_t step) {
+    VELOX_USER_CHECK_NE(step, 0, "step must not be zero");
+    VELOX_USER_CHECK(
+        step > 0 ? stop >= start : stop <= start,
+        "sequence stop value should be greater than or equal to start value if "
+        "step is greater than zero otherwise stop should be less than or equal to start")
+    VELOX_USER_CHECK_LE(
+        (stop - start) / step + 1,
```

Comment:
nit: to avoid computing this twice, consider returning this value from checkArguments.